### ISSUE TITLE
Queries are kept waiting even when there are available workers

### ISSUE THREAD

**Issue by shimpeko** [getredash/redash#1782]
### Issue Summary
I have 4 workers for adhoc/user queries. 4 workers means number of concurrency should be 4.
But sometimes we see less than 4 queries are in progress while there are waiting queries in queue on `/admin/queries/tasks` page.

I think this happens because of celery worker's prefetch function. A celery worker fetches 4 tasks(queries) at a time by default. If a worker fetches 4 tasks at once and 1st query is long running, rest of 3 queries will be kept waiting for long time even if other workers are available.

Prefetch is good for many short queries but not efficient with long running queries. So I think it should be configurable through `setting.py`.

About celery prefetch: http://docs.celeryproject.org/en/latest/userguide/optimizing.html#reserve-one-task-at-a-time

### Steps to Reproduce

1. Set up more than 2 workers
2. Run some queries

### Technical details:

- Redash Version: 1.0.3
- Browser/OS: Chrome
- How did you install Redash: Docker

**Comment by arikfr** [comment]
> I have 4 workers for adhoc/user queries.

Is it 4 Celery workers or one Celery worker with concurrency of 4 (`-c4`)?

**Comment by shimpeko** [comment]
4 celery workers and worker's concurrency is 1 (`-c1`).

**Comment by arikfr** [comment]
I'm using ECS too, but I use both multiple ECS services/tasks & Celery's concurrency, to better utilize memory allocations.

**Comment by arikfr** [comment]
But anyway, this explains why you might experience this. I'm just not sure that the solution in #1783 is the right direction :\

**Comment by nocode99** [comment]
@arikfr Can you post some example celery configurations you use in ECS with rough estimations on your workloads?  We are currently testing the latest version and running the workers in Docker.  We provisioned 3 containers with concurrency set to 2 (so a max of 6).  I also setup celery flower to monitor and we are seeing inconsistent behaviors on when workers pick up tasks.  Sometimes, we'll have 4 queries running while the rest are in queue.  Then, we saw 6 queries running simultaneously, but after a bunch of tasks finished, the # of simultaneous queries dropped to 4 while we still had queries in queue.  Flower showed that the 3rd worker had no queries running, but did not pick up any jobs.

We are planning to put the celery workers in ECS but our testing has been very inconsistent.  On average, our queries take anywhere between seconds to ~10 min.  We primarily use this towards our Redshift cluster where we allow up to 10 simultaneous queries.

**Comment by vbajaria** [comment]
@arikfr I am facing a similar problem in our re:dash setup. I have tried playing around with concurrency but nothing helps.

My queries are stuck in "Waiting" state forever and only a restart of the celery workers help them move forward. Most of the queries take a few seconds to run and return < 100 rows.

I am on 1.0.1+b2833

**Comment by nocode99** [comment]
So far, I can confirm the setting for the **CELERYD_PREFETCH_MULTIPLIER** in https://github.com/getredash/redash/pull/1783 has relieved queries in queue issue.  By default, the setting is set to 4, but for long running queries, the setting should be set to 1 according to celery's documentation.

Let's say you leave the default celery concurrency and spin up 4 more workers (x2 concurrency).  Redash should have 5 workers (10 queues) available.  If you submit 10 long running queries, only 3 workers may be used up while the other 2 sit idle.  4 would get sent to one worker, another 4 to the 2nd worker and the remaining 2 to a 3rd worker.  4 queries would be in queue since the first 2 workers can only handle 2 concurrent jobs.

I'm not so sure about the ACK setting in the PR, but the **CELERYD_PREFETCH_MULTIPLIER** would be a welcome addition (as well as the autoscale parameter).

**Comment by vbajaria** [comment]
@nocode99 thanks for the feedback. Is there a easy way to set the config without applying #1783 ?

While I agree that the example you give above is valid and the setting should alleviate it, the behavior that I am observing is weird.

My scheduled queries run fine. It's the adhoc queries that we run which get stuck. Currently I have only 1 query in "Waiting" state and for some reason celery is not picking up that task for execution. Have you ever noticed this behavior ?

**Comment by nlutterman-openmesh** [comment]
Just wanted to note that this issue is still occurring with version `4.0.1.b4038`.

I've increased the number of workers, tried applying #1783.  It's only resolved when I run:

    supervisorctl stop redash_celery && redis-cli flushall && supervisorctl start redash_celery

But that's only a temporary fix.  After running a few queries (5-6 small ones) the behavior starts over.
As a short-time solution, I'm going to implement a cron which runs the above command, but it appears to be an incorrect one, at least to me.

**Comment by nlutterman-openmesh** [comment]
I have discovered the source of my own issues. For the sake of posterity I'll include a full description here.

Celery has a [strict limit of 4 seconds](http://docs.celeryproject.org/en/latest/userguide/signals.html?highlight=worker_process_init#worker-process-init) for workers which use the `worker_process_init` signal.

Redash utilizes this signal for its workers.  Redash also instantiates a new instance of the application upon launching of a new worker where this signal is used.

In my case, the worker was taking longer than 4 seconds to initialize, and as `init_celery_flask_app()` is decorated with the `worker_process_init` signal in the file `redash/worker.py`, celery was destroying the new process. `supervisord` was configured to restart the process indefinitely, so celery would kill it, supervisord would restart it, and this would repeat until the process would launch in under 4 seconds.

The result of this, is that it would just appear (from the front-end) that the query was just taking anywhere from 5 seconds to 10 minutes for the task runner to run.  Queries would eventually run if left in the queue for long enough, as Redash would eventually initialize in worker form in under 4 seconds (after probably a few hundred tries).

I have not yet determined the cause of it taking longer than 4 seconds to start up, my first guess is opening connections to database servers.

My very short-term fix is to manually adjust celery's `PROC_ALIVE_TIMEOUT` constant to allow for Redash to complete its initialization. There _no_ configuration option for this.

In `redash/worker.py` I added the following:

```python
...
### Original ###
from celery import Celery
from celery.schedules import crontab
from celery.signals import worker_process_init

### Added the following: ###
from celery.concurrency import asynpool
asynpool.PROC_ALIVE_TIMEOUT = 10.0 #set this long enough
```

From examining the logs I saw that this allowed the worker to finish starting and actually execute the queries that were in the task queue.

I'll leave another comment when I figure out what is causing the slow startup time for the worker.

**Comment by arikfr** [comment]
@nlutterman-openmesh thank you for sharing!

**Comment by arikfr** [comment]
This will be resolved in #4092, because we will get rid of Celery's prefetching behavior.
